tchannel/peer_heap.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import absolute_import
+
+from .container import heap
+from .container.heap import HeapOperation
+
+
+class PeerHeap(HeapOperation):
+    def __init__(self):
+        self.peers = []
+
+    def size(self):
+        return len(self.peers)
+
+    def lt(self, i, j):
+        return self.peers[i].score < self.peers[j].score
+
+    def push(self, x):
+        x.index = len(self.peers)
+        self.peers.append(x)
+
+    def pop(self):
+        n = len(self.peers)
+        item = self.peers[n-1]
+        item.index = -1
+        del self.peers[n - 1]
+        return item
+
+    def swap(self, i, j):
+        self.peers[i], self.peers[j] = self.peers[j], self.peers[i]
+        self.peers[i].index = i
+        self.peers[j].index = j
+
+    def update_peer(self, peer):
+        """Update the peer's position in the heap after peer's score changed"""
+        heap.fix(self, peer.index)
+
+    def pop_peer(self):
+        """Pop the top peer from the heap
+
+        :return
+            return the top peer and remove it from the heap if heap is not
+            empty.Otherwise return None.
+        """
+        if len(self.peers) == 0:
+            return None
+        return heap.pop(self)
```

Comment:
So the public interface of the `PeerHeap` class is: `{push,pop,peek,update}_peer`, right? The other methods are for the `HeapOperation`. You don't need to fix this right now (add a TODO) but this could be two separate classes.  One that's the actual min-heap and the other the public interface for PeerHeap that wraps the min-heap based on the heap.py.